File: codewars/Snail/Snail.py
def snail(array):
    if len(array) == 0:
        return []
    if len(array) == 1:
        return array[0]
    
    top = array[0][:-1]
    right = [a[-1] for a in array]
    bottom = array[-1][:-1][::-1]
    left = [b[0] for b in array[1:-1]][::-1]
    inner = [c[1:-1] for c in array[1:-1]]
    
    return top + right + bottom + left + snail(inner)

File: codewars/Snail/snail.js
function snail(matrix) {
    if (matrix.length === 0) return [];
    const total = [];
    const rows = matrix.length;
    const cols = matrix[0].length;
    let top = 0, bottom = rows - 1;
    let left = 0, right = cols - 1;

    while (top <= bottom && left <= right) {
        for (let i = left; i <= right; i++) {
            total.push(matrix[top][i]);
        }
        top++;
        for (let i = top; i <= bottom; i++) {
            total.push(matrix[i][right]);
        }
        right--;
        if (top <= bottom) {
            for (let i = right; i >= left; i--) {
                total.push(matrix[bottom][i]);
            }
            bottom--;
        }

        if (left <= right) {
            for (let i = bottom; i >= top; i--) {
                total.push(matrix[i][left]);
            }
            left++;
        }
    }
    return total;
}